### PR TITLE
Implementation of the Ubuntu 20.04 ARM64 End-To-End Test Pipeline

### PR DESCRIPTION
**Pull request by nihemstr**
- Added the E2E Automated Test Pipeline for Ubuntu 20.04 ARM64
- Reconfigured the templates for adding new scenarios that allows for the parallelization of test setup/running

### CODE REVIEW

File: azurepipelines/e2e_test/scenarios/ubuntu-18.04-arm64/vm_setup/setup.sh [hunk 1]
```diff
@@ -164,6 +164,9 @@ function test_shutdown_service() {
 }
 
 configure_apt_repository
+
```
Comment (JeffMill) [inline azurepipelines/e2e_test/scenarios/ubuntu-18.04-arm64/vm_setup/setup.sh:167]: why was this moved up?

File: azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/Bundle-update.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
+# -------------------------------------------------------------------------
```
Comment (JeffMill) [inline azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/Bundle-update.py:1]: you're adding a bunch of files in /ubuntu-20.04-arm64, yet we have a /ubuntu-18.04-arm64 folder.  Is there really no/little overlap that would avoid a bunch of new files to be added?

File: azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/devicesetup.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import json
+import sys
+
+# Note: the intention is that this script is called like:
+# python ./<scenario-name>/vm-setup/devicesetup.py
+sys.path.append('./scenarios/')
```
Comment (JeffMill) [inline azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/devicesetup.py:11]: adding to the path seems like a code smell.  is there no better way to accomplish what you need to do?

File: azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/devicesetup.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+from testingtoolkit import DeviceUpdateTestHelper,DuAutomatedTestConfigurationManager
+
+# Note: the intention is that this script is called like:
+# python ./scenarios/<scenario-name>/<test-script-name>.py
+sys.path.append('./scenarios/ubuntu-20.04-arm64/')
```
Comment (JeffMill) [inline azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/devicesetup.py:16]: as above.

File: azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/devicesetup.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        jsonFile.write(configJson)
+
+
+if __name__ == '__main__':
+    main()
+
+def print_error(msg):
+    print(msg, file=sys.stderr)
+
+#
```
Comment (JeffMill) [inline azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/devicesetup.py:81]: seems like this should be in the file header (top of file), not at bottom.

File: azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/setup.sh [hunk 6]
```diff
@@ -0,0 +1,190 @@
+#!/bin/bash
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+# Setup Script for Test: Ubuntu-18.04-amd64-APT-deployment
```
Comment (JeffMill) [inline azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/setup.sh:7]: 18.04?
amd64?
APT?

File: azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/setup.sh [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+function register_extensions() {
+    mkdir ~/adu_srcs/
+
+    tar -xf ./testsetup/adu_srcs_repo.tar.gz -C ~/adu_srcs/
+
+    sudo mkdir -p ~/demo/demo-devices/contoso-devices
+
+    cd ~/adu_srcs/src/extensions/component_enumerators/examples/contoso_component_enumerator/demo || exit
+
+    chmod 755 ./tools/reset-demo-components.sh
```
Comment (JeffMill) [inline azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/setup.sh:67]: u+x ?

File: azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/setup.sh [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+    #copy components-inventory.json and adds it to the /usr/local/contoso-devices folder
+    sudo cp -a ./demo-devices/contoso-devices/. ~/demo/demo-devices/contoso-devices/
```
Comment (JeffMill) [inline azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/setup.sh:70]: why is sudo needed?

File: azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/setup.sh [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+    tar -xf ./testsetup/adu_srcs_repo.tar.gz -C ~/adu_srcs/
+
+    sudo mkdir -p ~/demo/demo-devices/contoso-devices
+
+    cd ~/adu_srcs/src/extensions/component_enumerators/examples/contoso_component_enumerator/demo || exit
+
+    chmod 755 ./tools/reset-demo-components.sh
+
+    #copy components-inventory.json and adds it to the /usr/local/contoso-devices folder
```
Comment (JeffMill) [inline azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/setup.sh:69]: huh?

File: azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/setup.sh [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    status_output=$(systemctl status deviceupdate-agent.service)
+
+    main_pid=$(echo "$status_output" | awk '/Main PID/ {print $3}')
+
+    # Check if the user and group for the process are "adu"
+    process_user=$(stat -c %u "/proc/$main_pid")
+    process_group=$(stat -c %g "/proc/$main_pid")
+
+    if [ "$process_user" -eq "$adu_uid" ] && [ "$process_group" -eq "$adu_gid" ]; then
+        echo "User and group for AducIotAgent service are adu:adu."
```
Comment (JeffMill) [inline azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/setup.sh:96]: nit: this trace not needed.

**Review by JeffMill — CHANGES_REQUESTED** [review]

File: azurepipelines/e2e_test/scenarios/ubuntu-18.04-arm64/vm_setup/setup.sh (hunk 1 above)
Comment (nihemstr) [reply azurepipelines/e2e_test/scenarios/ubuntu-18.04-arm64/vm_setup/setup.sh:167]: We need to have the APT repo configured before we do anything else so we know the APT repo on the machine has been configured with packages.microsoft.com

Comment (nihemstr) [reply azurepipelines/e2e_test/scenarios/ubuntu-18.04-arm64/vm_setup/setup.sh:167]: We need to have the APT repo configured before we do anything else so we know the APT repo on the machine has been configured with packages.microsoft.com

File: azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/Bundle-update.py (hunk 2 above)
Comment (nihemstr) [reply azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/Bundle-update.py:1]: It has to do with the original structure of the scenarios. When we began this we expected the end-to-end scenarios to require much different code. 

We're in desperate need of a refactoring tracked here: https://microsoft.visualstudio.com/OS/_queries/edit/44133514/?triage=true

so we can remove the redundancy. 

My plan is to roll that work together with the new provisioning work for the test scenarios that needs to be done this quarter. 

It's not something that can be addressed as apart of this PR but rather a change in architecture that MUST be made.

File: azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/devicesetup.py (hunk 3 above)
Comment (nihemstr) [reply azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/devicesetup.py:11]: Not for code you're not installing as a module. 

Another method we could use is creating a *.wheel file for the toolkit and then install it as apart of the requirements.txt file.

File: azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/devicesetup.py (hunk 4 above)
Comment (nihemstr) [reply azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/devicesetup.py:16]: It looks ugly and it is. Python projects like to be flat but it's needed for the way we're using it here. Python runs as a command-line tool so it bows to those values.


This is a good candidate for code quality improvements. Tracked here: https://microsoft.visualstudio.com/OS/_queries/edit/43984595/?triage=true&workitem=44136731

File: azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/setup.sh (hunk 6 above)
Comment (nihemstr) [reply azurepipelines/e2e_test/scenarios/ubuntu-20.04-arm64/vm_setup/setup.sh:7]: bad copy.

**Review by shiyi-peng — APPROVED** [review]
LGTM! Perfect

**Review by Nox-MSFT — APPROVED** [review]

**Review by JeffMill — APPROVED** [review]
